File: javascript/linked-list.js
class Node {
  constructor(value) {
    this.value = value
    this.next = null;
  }
}

export class LinkedList {
  constructor() {
    this.first = null;
    this.length = 0;
  }

  push(val) {
    let newNode = new Node(val);
    if (this.length === 0) {
      this.first = newNode;
    } else {
      let node = this.first;
      while (node.next) {
        node = node.next;
      }
      node.next = newNode;
    }
    this.length++;
    return this;
  }

  pop() {
    if (this.length === 0) {
      return null;
    } else if (this.length === 1) {
      let val = this.first.value;
      this.first = null;
      this.length--;
      return val;
    } else {
      let node = this.first;
      let prev = null;
      while (node.next) {
        prev = node;
        node = node.next
      }
      let val = node.value;
      prev.next = null;
      this.length--;
      return val;
    }
  }

  shift() {
    if (this.length === 0) {
      return null;
    } else {
      let { next, value } = this.first;
      this.first = next;
      this.length--;
      return value;
    }
  }

  unshift(val) {
    let newNode = new Node(val);
    if (this.length === 0) {
      this.first = newNode;
    } else {
      newNode.next = this.first;
      this.first = newNode;
    }
    this.length++;
    return this;
  }

  delete(val) {
    if (this.length === 0) return false;
    if (this.length === 1) {
      if (this.first.value === val) {
        this.first = null;
        this.length--;
        return true;
      }
    } else if (this.length === 2) {
      if (this.first.value === val) {
        this.first = this.first.next;
        this.length--;
        return true;
      }
      if (this.first.next.value === val) {
        this.first.next = null;
        this.length--;
        return true;
      }
    } else {
      let prev = null;
      let node = this.first;
      while (node.value !== val || node.next === null) {
        prev = node;
        node = node.next;
      }
      prev.next = node.next;
      this.length--;
      return true;
    }
    return false;
  }

  count() {
    return this.length;
  }
}

File: elixir/linked_list.ex
defmodule LinkedList do
  @opaque t :: tuple()

  @doc """
  Construct a new LinkedList
  """
  @spec new() :: t
  def new() do
    {}
  end

  @doc """
  Push an item onto a LinkedList
  """
  @spec push(t, any()) :: t
  def push(list, elem), do: {elem, list}

  @doc """
  Counts the number of elements in a LinkedList
  """
  @spec count(t) :: non_neg_integer()
  def count(list), do: count(list, 0)
  defp count(nil, acc), do: acc
  defp count({}, acc), do: acc
  defp count({_v, next}, acc), do: count(next, acc + 1)

  @doc """
  Determine if a LinkedList is empty
  """
  @spec empty?(t) :: boolean()
  def empty?(nil), do: true
  def empty?({}), do: true
  def empty?(_), do: false

  @doc """
  Get the value of a head of the LinkedList
  """
  @spec peek(t) :: {:ok, any()} | {:error, :empty_list}
  def peek(list) do
    if empty?(list) do
      {:error, :empty_list}
    else
      {first, _} = list
      {:ok, first}
    end
  end

  @doc """
  Get tail of a LinkedList
  """
  @spec tail(t) :: {:ok, t} | {:error, :empty_list}
  def tail(list) do
    if empty?(list) do
      {:error, :empty_list}
    else
      {_v, next} = list
      {:ok, next}
    end
  end

  @doc """
  Remove the head from a LinkedList
  """
  @spec pop(t) :: {:ok, any(), t} | {:error, :empty_list}
  def pop(list) do
    if empty?(list) do
      {:error, :empty_list}
    else
      {head, tail} = list
      {:ok, head, tail}
    end
  end

  @doc """
  Construct a LinkedList from a stdlib List
  """
  @spec from_list(list()) :: t
  def from_list([]), do: new()
  def from_list([x | xs]), do: push(from_list(xs), x)

  @doc """
  Construct a stdlib List LinkedList from a LinkedList
  """
  @spec to_list(t) :: list()
  def to_list(list) do
    if empty?(list) do
      []
    else
      {v, next} = list
      [v | to_list(next)]
    end
  end

  @doc """
  Reverse a LinkedList
  """
  @spec reverse(t) :: t
  def reverse(list), do: reverse(list, new())
  defp reverse(list, acc) do
    if empty?(list) do
      acc
    else
      {v, next} = list
      reverse(next, push(acc, v))
    end
  end
end
